television/utils/image.rs

```diff
@@ -0,0 +1,255 @@
+use image::imageops::FilterType;
+use image::{DynamicImage, GenericImageView, Pixel, Rgba};
+use ratatui::buffer::{Buffer, Cell};
+use ratatui::layout::{Position, Rect};
+use ratatui::prelude::Color;
+use ratatui::widgets::{Block, Widget};
+use std::fmt::Debug;
+use std::hash::{Hash, Hasher};
+use std::sync::{Arc, Mutex};
+
+static PIXEL_STRING: &str = "▀";
+const FILTER_TYPE: FilterType = FilterType::Lanczos3;
+
+// use to reduce the size of the image before storing it
+const CACHED_WIDTH: u32 = 128;
+const CACHED_HEIGHT: u32 = 128;
+
+const GRAY: Rgba<u8> = Rgba([242, 242, 242, 255]);
+const WHITE: Rgba<u8> = Rgba([255, 255, 255, 255]);
+
+pub struct ImagePreviewWidget<'a> {
+    block: Block<'a>,
+    displayed_image: Arc<Mutex<Option<DisplayedImage>>>,
+}
+impl<'a> ImagePreviewWidget<'a> {
+    pub fn new(
+        displayed_image: Arc<Mutex<Option<DisplayedImage>>>,
+        block: ratatui::widgets::Block<'a>,
+    ) -> ImagePreviewWidget<'a> {
+        ImagePreviewWidget {
+            block,
+            displayed_image,
+        }
+    }
+}
+impl Widget for ImagePreviewWidget<'_> {
+    fn render(self, area: Rect, buf: &mut Buffer)
+    where
+        Self: Sized,
+    {
+        // I also put the render of the block to be similar has the preview paragraph, but I believe they are both useless since the block is already rendered
+        let inner = self.block.inner(area);
+        self.block.render(area, buf);
+        self.displayed_image
+            .lock()
+            .unwrap()
+            .as_ref()
+            .unwrap()
+            .render(inner, buf);
+    }
+}
+
+#[derive(Default)]
+pub struct DisplayedImage {
+    pub area: Rect,
+    cells: Vec<Vec<Cell>>,
+}
+
+impl Widget for &DisplayedImage {
+    fn render(self, area: Rect, buf: &mut Buffer) {
+        let height = self.cells.len();
+        if height == 0 {
+            return;
+        }
+        let width = self.cells[0].len();
+        // offset of the left top corner where the image is centered
+        let x_offset = (usize::from(area.width + 2 * area.x) - width) / 2;
+        let y_offset = (usize::from(area.height + 2 * area.y) - height) / 2;
+
+        for (y, row) in self.cells.iter().enumerate() {
+            for (x, cell) in row.iter().enumerate() {
+                if let Some(buf_cell) = buf.cell_mut(Position::new(
+                    u16::try_from(x_offset + x).unwrap_or(u16::MAX),
+                    u16::try_from(y_offset + y).unwrap_or(u16::MAX),
+                )) {
+                    *buf_cell = cell.clone();
+                }
+            }
+        }
+    }
+}
+#[derive(Clone)]
+pub struct CachedImageData {
+    image: DynamicImage,
+    inner_cache: Arc<Mutex<Option<DisplayedImage>>>,
```

Comment:
```suggestion
    /// A potentially downscaled version of the original image
    image: DynamicImage,
    /// An inner cache destined to cache the final displayed image which might have
    /// been resized again to fit to the screen's dimensions
    inner_cache: Arc<Mutex<Option<DisplayedImage>>>,
```
just to make sure I understand this correctly